In [107]:
import math
import random

def generate_data(num_trials):
    trials = []
    for _ in range(num_trials):
        count = 0
        while True:
            count += 1
            if random.randint(1, 125) == 1:
                break
        trials.append(count)
    return trials

x_data = generate_data(1000)

def apply_formula(x):
    # 여기에 당신의 공식을 적용합니다.
    # 예: y = x * 2
    y = round(1-math.exp(-x/125),4)
    return y

def generate_y_list(x_list):
    y_list = []
    for x in x_list:
        y = apply_formula(x)
        y_list.append(y)
    return y_list

# Y 데이터 리스트 생성
y_data = generate_y_list(x_data)

print("X 데이터:", x_data)
print("Y 데이터:", y_data)

X 데이터: [33, 300, 100, 52, 124, 48, 83, 20, 7, 238, 499, 29, 479, 101, 55, 182, 230, 146, 165, 54, 38, 23, 349, 168, 10, 382, 484, 110, 19, 182, 71, 58, 424, 9, 129, 19, 29, 39, 179, 48, 122, 91, 10, 33, 239, 427, 173, 176, 330, 20, 9, 40, 341, 45, 184, 95, 70, 164, 228, 3, 31, 132, 296, 11, 109, 7, 102, 64, 92, 92, 159, 41, 93, 139, 79, 27, 152, 9, 7, 91, 23, 21, 153, 156, 231, 72, 77, 27, 344, 13, 239, 311, 497, 67, 3, 356, 338, 46, 37, 17, 197, 257, 375, 145, 34, 47, 32, 24, 56, 113, 13, 48, 38, 22, 137, 203, 6, 16, 131, 397, 184, 314, 21, 4, 136, 115, 115, 3, 110, 131, 437, 23, 114, 78, 362, 98, 75, 27, 16, 47, 7, 30, 152, 21, 44, 12, 86, 85, 117, 131, 39, 30, 215, 235, 63, 227, 236, 432, 120, 16, 147, 55, 35, 3, 39, 87, 89, 127, 41, 32, 74, 250, 196, 137, 2, 119, 456, 21, 150, 95, 69, 35, 173, 7, 369, 142, 18, 219, 83, 84, 33, 34, 132, 2, 31, 42, 75, 84, 233, 124, 272, 90, 137, 89, 73, 19, 62, 239, 62, 593, 23, 88, 51, 32, 113, 117, 47, 187, 234, 156, 79, 5, 366, 571, 510, 69, 158,

# 1단계 생각( 처음 시도 오류들 발견됨)

# Gaussian process 이용하려함

In [13]:
!pip install GPy

In [110]:
import GPy
import numpy as np

# 리스트를 한 칸씩 뒤로 미루고, 마지막 요소는 제거
y_data = x_data[:-1]
y_data.insert(0, 0)  # 리스트 시작 부분에 0 추가

# 예시 데이터: X는 시도 횟수, Y는 관측된 값
X = np.array(x_data).reshape(-1, 1)
Y = np.array(y_data).reshape(-1, 1)

# Matérn 커널 사용하여 가우시안 프로세스 회귀 모델 생성
kernel = GPy.kern.Matern32(input_dim=1)
model = GPy.models.GPRegression(X, Y, kernel)

# 모델 최적화
model.optimize()

# 새로운 데이터에 대한 예측
new_X = np.array([[146, 35, 86, 162, 6, 31, 386]])  # 새로운 입력 값
predicted_Y, variance = model.predict(new_X)

print("예측된 값:", predicted_Y)
print("예측의 불확실성:", variance)

# 예측 불확실성(분산) 보니 아주 개판이다. 불확실성이 엄청큼

예측된 값: [[125.88509367]]
예측의 불확실성: [[15456.26135939]]


# RNN 이용

이렇게하면 문제가 쫙 학습하고 숫자하나보고 나음올 숫자를 맞추는것이기때문에
숫자 배열을 통해 다음올 숫자를 고려하지는 못한다. 그러나 단순히 임의 숫자뒤에 어떤숫자가 올지 예측해준다. 이를 통해 일정 특징을 발견할수도있다. 아래 rnn 모델 문법적오류로 실행은 안되지만 위에 가우시안프로세스 보니까 결과 뻔해서 안해본다. 애초에 말이안되는 접근이다 처음 두모델(가우시안, rnn)모델은

In [30]:
!pip install tensorflow

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.7/199.7 MB 3.8 MB/s eta 0:00:0000:0100:02
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.5/9.5 MB 4.7 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.5/65.5 kB 5.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 9.1 MB/s eta 0:00:0000:0100:01m
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 14.9 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.6/20.6 MB 9.8 MB/s eta 0:00:00:00:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 kB 5.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 440.7/440.7 kB 1.2 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 3.5 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 3.7 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 183.3/183.3 kB 3.4 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [109]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense

original_list = [x_data]
new_element = 0  # 리스트 시작 부분에 추가할 새로운 요소

# 리스트를 한 칸씩 뒤로 미루고, 새로운 요소를 추가
shifted_x_data = [new_element] + original_list[:-1]

# 예시 데이터셋 (실제 데이터로 교체 필요)
data = [x_data]  # 입력 시퀀스
next_number = [shifted_x_data]  # 예측해야 할 다음 숫자

# 데이터 전처리
X = np.array(data).reshape((len(data), 1, 1))
Y = np.array(next_number).reshape((len(next_number), 1))

# RNN 모델 구축
model = Sequential([
    SimpleRNN(50, activation='relu', input_shape=(1, 1)),
    Dense(1)
])
model.compile(optimizer='adam', loss='mse')

# 모델 학습
model.fit(X, Y, epochs=200, verbose=0)

# 새로운 데이터에 대한 예측
new_data = np.array([21]).reshape((1, 1, 1))
predicted_number = model.predict(new_data)
print("예측된 다음 숫자:", predicted_number[0][0])


ValueError: cannot reshape array of size 1000 into shape (1,1,1)

# 2단계 생각
내가 직접 포아송분포 이용해서 확률을 계산해서 학습시켜서 하려했는데 녹녹치 않아서
그냥 숫자들을 리스트로주고 스스로 확률계산하여 숫자배열의 특성을 밝히는 모델을 먼저 찾아보자 찾아보는데 딱히 안보여서 내가 직접 슈도코드 만들어보고(수학적으로) 코드로 만들어서 모델을 만들어본다.

# LSTM 이용시도
10개씩 쪼개어서 그다음 나오는 숫자를 타겟으로 하여 학습시키는 것도 가능할것같다. 근데 이렇게하면 10개가 채워지기 전에는 예측을 해주지 못한다. 하지만 10개가 넘어가고 부터는 계속 예측가능하다.

In [62]:
from keras.models import Sequential
from keras.layers import LSTM, Dense
import numpy as np

# 데이터셋 생성 함수
def create_dataset(numbers, n_steps):
    X, y = [], []
    for i in range(len(numbers)):
        end_ix = i + n_steps
        if end_ix > len(numbers)-1:
            break
        seq_x, seq_y = numbers[i:end_ix], numbers[end_ix]
        X.append(seq_x)
        y.append(seq_y)
    return np.array(X), np.array(y)

# 숫자 시퀀스 정의
numbers = x_data
n_steps = 8 # 학습하고싶은 숫자묶음의 크기(발걸음)을 설정합니다.

# 데이터셋 생성
X, y = create_dataset(numbers, n_steps)

# LSTM 모델 정의
model = Sequential()
model.add(LSTM(50, activation='relu', input_shape=(n_steps, 1)))
model.add(Dense(1))
model.compile(optimizer='adam', loss='mse')

# 모델 훈련을 위해 X를 재구성
X = X.reshape((X.shape[0], X.shape[1], 1))

# 모델 훈련
model.fit(X, y, epochs=200, verbose=0)

# 예측을 위한 새로운 데이터 포인트
x_input = np.array([49, 233, 79, 141, 93, 199, 35, 3])
x_input = x_input.reshape((1, n_steps, 1))

# 다음 숫자 예측
yhat = model.predict(x_input, verbose=1) # verbose 설정에따라 학습상황을 아려줍니다.
yhat[0][0]  # 예측된 다음 숫자 출력

1/1 [==============================] - 0s 75ms/step


128.75371

# LSTM 숫자 우연인지 몇번 정확히 맞췄다.
데이터 세트의 크기를 키우며 어디서 맞추는지 실험해본다. 1.스텝의 크기 2.데이터세트의 크기 조절해볼것 세트 크기 (1000개:120예측 10000:90예측 2000:112예측) 1000개에서 더 근사하게 맞춘다.
손좀봐보자
정확히 맞춘것은 데이터세트 1000개 숫자배열 크기가 8일때였다
###같은 조건으로 반복하여도 예측숫자는 다르게 출력된다.
### 계속해보니  확률적으로 근사하여서 맞추는것같긴하다
우연이 아니었다 1000개 8개일때 적중하는 숫자있다. 어느정도 맞추는것같다 모델 조금더 손보자 그리고 (나중에 내가 모델 만들어보기도 하자)
for문으로 5번정도 시켜서 숫자보기
실험 (159 130 170 109 128)